lib/stellar/sep10.rb

```diff
@@ -0,0 +1,364 @@
+      server: Stellar::KeyPair
+    ] => nil)
+    # DEPRECATED: Use verify_challenge_tx_signers instead.
+    # This function does not support multiple client signatures.
+    #
+    # Verifies if a transaction is a valid per SEP-10 challenge transaction, if the validation 
+    # fails, an exception will be thrown.
+    #
+    # This function performs the following checks:
+    #   1. verify that transaction sequenceNumber is equal to zero;
+    #   2. verify that transaction source account is equal to the server's signing key;
+    #   3. verify that transaction has time bounds set, and that current time is between the minimum and maximum bounds;
+    #   4. verify that transaction contains a single Manage Data operation and it's source account is not null;
+    #   5. verify that transaction envelope has a correct signature by server's signing key;
+    #   6. verify that transaction envelope has a correct signature by the operation's source account;
+    #
+    # @param challenge_xdr [String] SEP0010 transaction challenge transaction in base64.
+    # @param server [Stellar::KeyPair] keypair for server's account.
+    #
+    # Raises a InvalidSep10ChallengeError if the validation fails
+    def self.verify_challenge_tx(
+      challenge_xdr: String, server: Stellar::KeyPair
+    )
+      transaction_envelope, client_address = read_challenge_tx(
+          challenge_xdr: challenge_xdr, server: server
+      )
+      client_keypair = Stellar::KeyPair.from_address(client_address)
+      if !verify_tx_signed_by(tx_envelope: transaction_envelope, keypair: client_keypair)
+        raise InvalidSep10ChallengeError.new(
+            "Transaction not signed by client: %s" % [client_keypair.address]
+        )
+      end
+    end
+
+    Contract(C::KeywordArgs[
+      tx_envelope: Stellar::TransactionEnvelope, 
+      signers: SetOf[::Hash]
+    ] => SetOf[::Hash])
+    # Verifies every signer passed matches a signature on the transaction exactly once,
+    # returning a list of unique signers that were found to have signed the transaction.
+    #
+    # @param tx_envelope [Stellar::TransactionEnvelope] SEP0010 transaction challenge transaction envelope.
+    # @param signers [SetOf[::Hash]] The signers of client account.
+    #
+    # @return [SetOf[::Hash]]
+    def self.verify_tx_signatures(
+      tx_envelope:,
+      signers:
+    )
+      signatures = tx_envelope.signatures
+      if signatures.empty?
+        raise InvalidSep10ChallengeError.new("Transaction has no signatures.")
+      end
+
+      signatures_used = Set.new
+      signers_found = Set.new
+      signers.each do |signer|
+        kp = Stellar::KeyPair.from_address(signer['key'])
+        tx_envelope.signatures.each_with_index do |sig, i|
+          if signatures_used.include?(i)
+            next
+          end
+          if sig.hint != kp.signature_hint
+            next
+          end
+          if verify_tx_signed_by(tx_envelope: tx_envelope, keypair: kp)
```

Comment:
`verify_tx_signed_by` hashes the transaction, and iterates over every signature again, so this is O(n2) because you're already inside a loop iterating over every signature. I don't think you need to call `verify_tx_signed_by` because you've already got the signer and the signature loaded here, you can just do the verify call on the signer. You can add one line to this function to calculate the hash further up and do it once.
```suggestion
          if kp.verify(sig.signature, tx_hash)
```